File: CS/WindowsApplication3/Custom Ribbon/MyRibbonItemViewInfoCalculator.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.Skins;
using DevExpress.XtraBars.Ribbon.ViewInfo;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Ribbon;
using DevExpress.Utils.Drawing;
using System.Drawing;
using DevExpress.Utils.Drawing.Animation;
using DevExpress.XtraBars.Ribbon.Drawing;
using DevExpress.Utils.Paint;
using DevExpress.XtraEditors.Drawing;

namespace DXSample {
    public class MyRibbonItemViewInfoCalculator : RibbonItemViewInfoCalculator
    {
        public MyRibbonItemViewInfoCalculator(BaseRibbonViewInfo viewInfo) : base(viewInfo) { }

        public static new void DrawItem(GraphicsCache cache, RibbonItemViewInfo itemInfo)
        {
            XtraAnimator.Current.DrawAnimationHelper(cache, itemInfo.Owner as ISupportXtraAnimation, itemInfo,
                new RibbonItemPainter(), new RibbonDrawInfo(itemInfo), null, itemInfo);
        }

        public override void DrawButton(DevExpress.Utils.Drawing.GraphicsCache cache, RibbonItemViewInfo item)
        {
            CheckItemViewInfo(cache, item);
            CustomDrawBarItemEventArgs args = new CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds,
                item.CaptionBounds, item.GlyphBounds, item.Item as BarItemLink, IsItemHotTracked(item));
            Ribbon.RaiseCustomDrawBarItemLink(args);
            if (!args.Handled)
                base.DrawButton(cache, item);
            else
                DrawSelection(cache, item);
        }

        public override void DrawDropDownButton(GraphicsCache cache, RibbonItemViewInfo item)
        {
            CheckItemViewInfo(cache, item);
            CustomDrawBarItemEventArgs args = new CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds,
                  item.CaptionBounds, item.GlyphBounds, item.Item as BarItemLink, IsItemHotTracked(item));
            Ribbon.RaiseCustomDrawBarItemLink(args);
            if (!args.Handled)
                base.DrawDropDownButton(cache, item);
            else
                DrawSelection(cache, item);
          
        }
        public override void DrawLargeDropDownButton(GraphicsCache cache, RibbonItemViewInfo item)
        {
            CheckItemViewInfo(cache, item);
            CustomDrawBarItemEventArgs args = new CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds,
                 item.CaptionBounds, item.GlyphBounds, item.Item as BarItemLink, IsItemHotTracked(item));
            Ribbon.RaiseCustomDrawBarItemLink(args);
            if (!args.Handled)
                base.DrawLargeDropDownButton(cache, item);
            else
                DrawSelection(cache, item);
        }

        public override void DrawEditItem(DevExpress.Utils.Drawing.GraphicsCache cache, RibbonItemViewInfo item)
        {
            CheckItemViewInfo(cache, item);
            CustomDrawBarItemEventArgs args = new CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds,
                 item.CaptionBounds, item.GlyphBounds, item.Item as BarItemLink, IsItemHotTracked(item));
            Ribbon.RaiseCustomDrawBarItemLink(args);
            if (!args.Handled)
                base.DrawEditItem(cache, item);
            else
            {
                RibbonEditItemViewInfo editItem = (RibbonEditItemViewInfo)item;
                if (editItem.EditBounds.IsEmpty) return;
                XPaint oldPaint = null;
                ControlGraphicsInfoArgs infoArgs = new ControlGraphicsInfoArgs(editItem.EditViewInfo, cache, editItem.EditBounds, ShouldDrawTextOnGlass(item));
                if (!ViewInfo.OwnerControl.Enabled)
                    infoArgs.ViewInfo.State = ObjectState.Disabled;
                editItem.EditPainter.Draw(infoArgs);
                if (oldPaint != null) cache.Paint = oldPaint;
                DrawSelection(cache, item);
            }
        }

        public override void DrawLargeButton(GraphicsCache cache, RibbonItemViewInfo item)
        {
            CheckItemViewInfo(cache, item);
            CustomDrawBarItemEventArgs args = new CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds,
                 item.CaptionBounds, item.GlyphBounds, item.Item as BarItemLink, IsItemHotTracked(item));
            Ribbon.RaiseCustomDrawBarItemLink(args);
            if (!args.Handled)
                base.DrawLargeButton(cache, item);
            else
                DrawSelection(cache, item);
        }

        protected virtual bool IsItemHotTracked(RibbonItemViewInfo item)
        {
            return ViewInfo.Manager.SelectionInfo.HighlightedLink == item.Item;
        }

        public new MyRibbonViewInfo ViewInfo
        {
            get { return base.ViewInfo as MyRibbonViewInfo; }
        }

        public MyRibbonControl Ribbon
        {
            get { return ViewInfo.Ribbon as MyRibbonControl; }
        }
    }
}

File: VB/WindowsApplication3/Custom Ribbon/MyRibbonItemViewInfoCalculator.vb
Imports DevExpress.XtraBars.Ribbon.ViewInfo
Imports DevExpress.XtraBars
Imports DevExpress.Utils.Drawing
Imports System.Drawing
Imports DevExpress.Utils.Drawing.Animation
Imports DevExpress.XtraBars.Ribbon.Drawing
Imports DevExpress.Utils.Paint
Imports DevExpress.XtraEditors.Drawing

Namespace DXSample

    Public Class MyRibbonItemViewInfoCalculator
        Inherits RibbonItemViewInfoCalculator

        Public Sub New(ByVal viewInfo As BaseRibbonViewInfo)
            MyBase.New(viewInfo)
        End Sub

        Public Shared Overloads Sub DrawItem(ByVal cache As GraphicsCache, ByVal itemInfo As RibbonItemViewInfo)
            Call XtraAnimator.Current.DrawAnimationHelper(cache, TryCast(itemInfo.Owner, ISupportXtraAnimation), itemInfo, New RibbonItemPainter(), New RibbonDrawInfo(itemInfo), Nothing, itemInfo)
        End Sub

        Public Overrides Sub DrawButton(ByVal cache As DevExpress.Utils.Drawing.GraphicsCache, ByVal item As RibbonItemViewInfo)
            CheckItemViewInfo(cache, item)
            Dim args As CustomDrawBarItemEventArgs = New CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds, item.CaptionBounds, item.GlyphBounds, TryCast(item.Item, BarItemLink), IsItemHotTracked(item))
            Ribbon.RaiseCustomDrawBarItemLink(args)
            If Not args.Handled Then
                MyBase.DrawButton(cache, item)
            Else
                DrawSelection(cache, item)
            End If
        End Sub

        Public Overrides Sub DrawDropDownButton(ByVal cache As GraphicsCache, ByVal item As RibbonItemViewInfo)
            CheckItemViewInfo(cache, item)
            Dim args As CustomDrawBarItemEventArgs = New CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds, item.CaptionBounds, item.GlyphBounds, TryCast(item.Item, BarItemLink), IsItemHotTracked(item))
            Ribbon.RaiseCustomDrawBarItemLink(args)
            If Not args.Handled Then
                MyBase.DrawDropDownButton(cache, item)
            Else
                DrawSelection(cache, item)
            End If
        End Sub

        Public Overrides Sub DrawLargeDropDownButton(ByVal cache As GraphicsCache, ByVal item As RibbonItemViewInfo)
            CheckItemViewInfo(cache, item)
            Dim args As CustomDrawBarItemEventArgs = New CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds, item.CaptionBounds, item.GlyphBounds, TryCast(item.Item, BarItemLink), IsItemHotTracked(item))
            Ribbon.RaiseCustomDrawBarItemLink(args)
            If Not args.Handled Then
                MyBase.DrawLargeDropDownButton(cache, item)
            Else
                DrawSelection(cache, item)
            End If
        End Sub

        Public Overrides Sub DrawEditItem(ByVal cache As DevExpress.Utils.Drawing.GraphicsCache, ByVal item As RibbonItemViewInfo)
            CheckItemViewInfo(cache, item)
            Dim args As CustomDrawBarItemEventArgs = New CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds, item.CaptionBounds, item.GlyphBounds, TryCast(item.Item, BarItemLink), IsItemHotTracked(item))
            Ribbon.RaiseCustomDrawBarItemLink(args)
            If Not args.Handled Then
                MyBase.DrawEditItem(cache, item)
            Else
                Dim editItem As RibbonEditItemViewInfo = CType(item, RibbonEditItemViewInfo)
                If editItem.EditBounds.IsEmpty Then Return
                Dim oldPaint As XPaint = Nothing
                Dim infoArgs As ControlGraphicsInfoArgs = New ControlGraphicsInfoArgs(editItem.EditViewInfo, cache, editItem.EditBounds, ShouldDrawTextOnGlass(item))
                If Not ViewInfo.OwnerControl.Enabled Then infoArgs.ViewInfo.State = ObjectState.Disabled
                editItem.EditPainter.Draw(infoArgs)
                If oldPaint IsNot Nothing Then cache.Paint = oldPaint
                DrawSelection(cache, item)
            End If
        End Sub

        Public Overrides Sub DrawLargeButton(ByVal cache As GraphicsCache, ByVal item As RibbonItemViewInfo)
            CheckItemViewInfo(cache, item)
            Dim args As CustomDrawBarItemEventArgs = New CustomDrawBarItemEventArgs(cache, item.Appearance, item.Bounds, item.CaptionBounds, item.GlyphBounds, TryCast(item.Item, BarItemLink), IsItemHotTracked(item))
            Ribbon.RaiseCustomDrawBarItemLink(args)
            If Not args.Handled Then
                MyBase.DrawLargeButton(cache, item)
            Else
                DrawSelection(cache, item)
            End If
        End Sub

        Protected Overridable Function IsItemHotTracked(ByVal item As RibbonItemViewInfo) As Boolean
            Return ViewInfo.Manager.SelectionInfo.HighlightedLink Is item.Item
        End Function

        Public Overloads ReadOnly Property ViewInfo As MyRibbonViewInfo
            Get
                Return TryCast(MyBase.ViewInfo, MyRibbonViewInfo)
            End Get
        End Property

        Public ReadOnly Property Ribbon As MyRibbonControl
            Get
                Return TryCast(ViewInfo.Ribbon, MyRibbonControl)
            End Get
        End Property
    End Class
End Namespace
